### ISSUE TITLE
master setup fails

### ISSUE THREAD

**Issue by jelmd** [ontohub/ontohub#1370]
When trying to setup ta.iws.cs.ovgu.de as the new master from master, precompiling assets fail:
```
/local/usr/ruby/versions/2.1.0/bin/ruby /local/home/ontohub/bin/rake assets:precompile:all RAILS_ENV=production RAILS_GROUPS=assets
rake aborted!
SettingsValidator::ValidationError: {:initializers__data_root=>["is not a directory"], :initializers__git_root=>["is not a directory"], :initializers__symlink_path=>["is not a directory"], :initializers__commits_path=>["is not a directory"], :yml__email=>["email adress must belong to the fully qualified domain name 'ta.iws.cs.ovgu.de'."], :yml__action_mailer__delivery_method=>["is not included in the list"], :yml__hets__executable_path=>["must be an executable file: /usr/bin/hets"]}
/local/home/ontohub/webapp/lib/settings_validator.rb:7:in `validate!'
/local/home/ontohub/webapp/config/application.rb:100:in `block in <class:Application>'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/lazy_load_hooks.rb:34:in `call'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/lazy_load_hooks.rb:34:in `execute_hook'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/lazy_load_hooks.rb:43:in `block in run_load_hooks'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/lazy_load_hooks.rb:42:in `each'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/lazy_load_hooks.rb:42:in `run_load_hooks'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/application/finisher.rb:59:in `block in <module:Finisher>'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:30:in `instance_exec'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:30:in `run'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:55:in `block in run_initializers'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:54:in `each'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/initializable.rb:54:in `run_initializers'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/application.rb:136:in `initialize!'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/railtie/configurable.rb:30:in `method_missing'
/local/home/ontohub/webapp/config/environment.rb:5:in `<top (required)>'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:251:in `require'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:251:in `block in require'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:236:in `load_dependency'
/local/home/ontohub/ruby/2.1.0/gems/activesupport-3.2.21/lib/active_support/dependencies.rb:251:in `require'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/application.rb:103:in `require_environment!'
/local/home/ontohub/ruby/2.1.0/gems/railties-3.2.21/lib/rails/application.rb:305:in `block (2 levels) in initialize_tasks'
/local/home/ontohub/ruby/2.1.0/gems/actionpack-3.2.21/lib/sprockets/assets.rake:93:in `block (2 levels) in <top (required)>'
/local/home/ontohub/ruby/2.1.0/gems/actionpack-3.2.21/lib/sprockets/assets.rake:60:in `block (3 levels) in <top (required)>'
Tasks: TOP => environment
(See full trace by running task with --trace)
rake aborted!
Command failed with status (1): [/local/usr/ruby/versions/2.1.0/bin/ruby /l...]
/local/home/ontohub/ruby/2.1.0/gems/actionpack-3.2.21/lib/sprockets/assets.rake:12:in `ruby_rake_task'
/local/home/ontohub/ruby/2.1.0/gems/actionpack-3.2.21/lib/sprockets/assets.rake:21:in `invoke_or_reboot_rake_task'
/local/home/ontohub/ruby/2.1.0/gems/actionpack-3.2.21/lib/sprockets/assets.rake:29:in `block (2 levels) in <top (required)>'
Tasks: TOP => assets:precompile
(See full trace by running task with --trace)
```

**Comment by eugenk** [comment]
This means your settings don't validate. Please check your `config/settings.local.yml`.

> :initializers__data_root=>["is not a directory"], :initializers__git_root=>["is not a directory"], :initializers__symlink_path=>["is not a directory"], :initializers__commits_path=>["is not a directory"]

Those mean, that you have not set `paths.data`, `paths.git_repositories`, `paths.symlinks` and `paths.commits` to actually existing directories.

> :yml__email=>["email adress must belong to the fully qualified domain name 'ta.iws.cs.ovgu.de'."]

The email address `email` must be of the form `something@ta.iws.cs.ovgu.de` if the hostname is `ta.iws.cs.ovgu.de`.

> :yml__action_mailer__delivery_method=>["is not included in the list"]

The `settings.yml` lists the allowed values:
>   # possible values for delivery_method (see ActionMailer documentation):
   # :smtp, :sendmail, :file, :test

And last:
> :yml__hets__executable_path=>["must be an executable file: /usr/bin/hets"]

`hets.executable_path` must be a path to an executable file.

**Comment by jelmd** [comment]
a) Hmm, wondering, why it is not possible to emit human readable error messages ...

b) pathes: why are there no reasonable defaults? Where is documented, what needs to be set?

c) I'm reluctant wrt. to all this yml__*- gues-what-it-is  messages. Print out which settings where need to be fixed, and than I can do it.

**Comment by eugenk** [comment]
> a) Hmm, wondering, why it is not possible to emit human readable error messages ...

There was too much to do and we're a *very* small team. Next step in settings validation is to print better readable error messages.

> b) pathes: why are there no reasonable defaults? Where is documented, what needs to be set?

Actually those defaults should be reasonable. The thing is that `rails_root/data` is not in the git repository. As far as I know, capistrano creates this directory at deploy time. @0robustus1, please correct me if I'm wrong.

> c) I'm reluctant wrt. to all this yml__*- gues-what-it-is  messages. Print out which settings where need to be fixed, and than I can do it.

See my answer to a). Also, I explained the messages in my first answer ;)
There should be no other validations failing. The system is to collect all invalid settings and print them together.

**Comment by jelmd** [comment]
a) a really good joke. 
b) everything else: plz. provide a working example.
Current state is: application doesn't work and the gobbledeegook it spits out isn't helpful at all resolving the problem.

**Comment by tillmo** [comment]
a) is not a joke at all, but reality. If you know good rails programmers in MD, please tell us.

**Comment by jelmd** [comment]
Oh, sorry, I assumed, that someone was bored to death and that's why started to implement something, which has obviously absolute no value. So wasting several weeks for a simple task, which takes an undergraduate probably less than 2 hours doesn't align with what I understand as "had no time". And if IIRC there was already an implementation, which would have taken how many time to integrate? Minutes? So please, stop kidding me!

**Comment by tillmo** [comment]
Please stop talking in this tome! Your implementation was not useful, because it did not integrate well in the Ontohub project and its coding guidelines. So someone else has to reimplement this. Obviously, this was only one little task among dozens others.

**Comment by eugenk** [comment]
As far as I remember, you want the data - especially the application-internal git repositories - to be located in `/data`. I suggest you do
```bash
mkdir -p /data/{repositories,git_daemon,commits}
```
and add to the `config/settings.local.yml`
```yml
paths:
  data: /data
  git_repositories: /data/repositories
  symlinks: /data/git_daemon
  commits: /data/commits
```
Note that - other than in the `config/settings.yml` - the paths begin with a slash.

--

Second, the path to Hets must be actually an executable file. Put the output of `which hets` into the `config/settings.local.yml`, for example:
```yml
hets:
  executable_path: /usr/local/bin/hets
```

--

Next, the email address is bound to the hostname. So if machine's hostname is `ta.iws.cs.ovgu.de` and you don't specify the hostname in the `config/settings.local.yml`, then you must choose an email address of that host in this config file:
```yml
email: noreply@ta.iws.cs.ovgu.de
```
Or, for example, specifying the hostname and email would look something like
```yml
hostname: ontohub.org
email: noreply@ontohub.org
```

--

Also, the mail method should be specified. I assume you want to use sendmail:
```yml
action_mailer:
  delivery_method: :sendmail
```
Possible values are `:smtp`, `:sendmail`, `:file` and `:test`. Note that they all begin with a colon.

--

Eventually the `config/settings.local.yml` should look something like this:
```yml
paths:
  data: /data
  git_repositories: /data/repositories
  symlinks: /data/git_daemon
  commits: /data/commits
hets:
  executable_path: /usr/local/bin/hets
email: noreply@ta.iws.cs.ovgu.de
action_mailer:
  delivery_method: :sendmail
```
assuming the paths exist.

----

I skimmed the current `post-install2.sh` of `td` and found some things worth noting. There were a few changes to the application and to the [productive deployment wiki page](https://github.com/ontohub/ontohub/blob/staging/doc/productive_deployment.md). I can't guarantee that this is everything (I'm not a server guy), but at least you have some more hints on what needs to be changed:


You place the `maintenance.txt` in the ontohub directory, but it belongs into the data directory:
```apache
	RewriteCond '"${OHOME}"'/maintenance.txt -f
```
Should be something like
```apache
	RewriteCond /data/maintenance.txt -f
```
if /data is your data directory (see above).

--

You have a function `postElasticsearch`. This should be complemented with what the [Securing Elasticsearch section](https://github.com/ontohub/ontohub/blob/staging/doc/productive_deployment.md#securing-elasticsearch) in our wiki says. This section is quite new.

-- 

You have a function `postGit`. It sets up the `cp_keys` file. You don't need some of the code anymore, because we have a rake task that reads the settings (including .local.yml) and hard-codes the paths into the c-source file and compiles it. You only need to call this task and then set the file permissions on the resulting binary. Read [the wiki section](https://github.com/ontohub/ontohub/blob/staging/doc/productive_deployment.md#changing-settings) for more details.

--

You have a function `postSolr`. Solr is not needed anymore. It was replaced by elasticsearch. On master, this happened with the last staging merge/the last release.

--

I hope that helps.
